crates/bevy_app/src/app.rs

```diff
@@ -92,7 +92,7 @@ impl Default for App {
         app.sub_apps.main.update_schedule = Some(Main.intern());
 
         #[cfg(feature = "bevy_reflect")]
-        app.init_resource::<AppTypeRegistry>();
+        app.insert_resource(AppTypeRegistry::new_with_derived_types());
```

Comment:
I think this entire feature, including this initial automatic registration should be behind a feature flag. The feature can be default since ideally this should almost always be used, but we want to use a feature so users who don't need this functionality (i.e. maybe they have no reliance on reflection) don't have to pay any penalties. 

Putting it behind a feature will also make it easier to roll back in the case that this causes unforeseen issues (which hopefully be caught via dogfooding, but might also be discovered in production).

Reply:
Yeah, I agree that there should probably a feature flag for this. However, I also remember there was some discussion about it and how this would break plugins that rely on the assumption that automatic reflect is enabled.

Reply:
Yeah plugins that rely on it will break if disabled, but it's much easier to have those broken plugins publish a release with the fix (or fork it) than for Bevy to do that.

Reply:
I agree. This has overhead (however small), so some users will want to disable it (especially in a `no_std` context!), and that shouldn't break plugins since, really, this a convenience for end-users. A _really_ nice convenience, but it doesn't unblock any particular feature or functionality you can't already do with manual registration.

Reply:
Put pretty much everything behind a feature gate. The only part that isn't is `no_auto_register` attribute.